### PR TITLE
ADBDEV-5772 [C] Extend list of types to predicate pushdown

### PR DESCRIPTION
**Pull request by bimboterminator1**
Extend the list of types for predicate pushdown

The recent changes to the pxffilters.c file in the external-table codebase and
in pxf_filter.c in the fdw codebase introduce several enhancements and
adjustments aimed at improving filter handling, particularly extending of
supported data types range.
Here is a summary of the key updates:

Array comparison support in the filters:
The operator map for pxf_supported_opr_op_expr now includes generic array
comparison operators like ARRAY_EQ_OP, enhancing the filter capabilities for
array types.  Only = , <>, IN, IS NULL operators are supported. If the array's
elements do not support any scalar operator, the error is thrown in order
to avoid extra work of pxf (otherwise GPDB will reject the result and throw an
error after the data has been processed and returned by the pxf).
A new oid list pxf_supported_array_types[] and a new function
supported_array_type(Oid type) are added to check if a given array type is
supported for filtering. It is used to process array as list constants
inside the opexpr_to_pxffilter() (OpExprToPxfFilter()) function.
pxf_serialize_filter_list() (PxfSerializeFilterList()) is updated to handle
scenarios where a filter operand is an attribute and the other is a list
constant.

List Constant Handling:
The function list_const_to_str() (ListConstToStr()) now includes an additional
bool parameter with_nulls, allowing it to handle NULL values in array constants
appropriately. For the case of IN operators (scalar_array_op_expr_to_pxffilter()
) the NULL value inside the enumeration is meaningless (the comparison with
NULLS can't be performed) and the list_const_to_str() is called with false
with_nulls argument. For the case of array comparison (like a =
ARRAY[TRUE, NULL]) the list_const_to_str() is called with true parameter
with_nulls.

BOOLARRAY type introduction:
New OID BOOLARRAYOID is defined as a new macro and is covered in #ifndef
because GPDB 6 does not have the definition for this type. 
The patch makes updates to pxf_supported_opr_scalar_array_op_expr with new entry
BooleanEqualOperator to handle the IN operator for BOOL type. The
list_const_to_str() (ListConstToStr()) also uses predefined string constants
to encode the bool values in the final filter representation.

Other types are handled by extending pxf_supported_opr_op_expr[],
pxf_supported_types[], pxf_supported_array_types[],
pxf_supported_opr_scalar_array_op_expr[] lists and corresponding switch
operators inside the scalar_const_to_str() and list_const_to_str() functions.

The following new data types have been added or extended for array types:

BYTEA, BYTEAARRAY;
FLOAT4, FLOAT8, FLOAT4ARRAY, FLOAT8ARRAY;
BPCHARARRAY;
VARCHARARRAYOID (here the modification of scalar_array_op_expr_to_pxffilter()
was needed due to RelabelType occurence in queries like SELECT * from
test_varchar where t in ('aaa'::varchar(10), 'bbb'::varchar(10));
DATEARRAY;
TIME, TIMEARRAY;
TIMESTAMPARRAY;
TIMESTAMPTZ, TIMESTAMPTZARRAY;
INTERVAL, INTERVALARRAYOID;
NUMERICARRAY;
UUID, UUIDARRAY;
JSONB, JSONBARRAY;
JSON, JSONARRAY (scalar operators and IN are not supported); 

In order to launch the regression tests the UserDataVerifyAccessor.java
has been extended to generate more fields of newly supported types.
Automation tests checkFilterPushdown were modified and extendted
to cover most of the newly added types.

----
### Example How to test with jdbc:
First of all, build and install pxf server
.
Use a text editor to uncomment the following line in the $PXF_BASE/conf/pxf-application.properties file and set the desired log level `pxf.log.level=debug`

With external table:
1. create jdbc profile in $PXF_BASE/conf/pxf-profiles.xml
<details>
  <summary>Ex</summary>

 ```

<?xml version='1.0' encoding='UTF-8'?>
<profiles>
  <profile>
    <name>JDBC</name>
    <description>jdbc</description>
    <plugins>
      <fragmenter>org.greenplum.pxf.plugins.jdbc.JdbcPartitionFragmenter</fragmenter>
      <accessor>org.greenplum.pxf.plugins.jdbc.JdbcAccessor</accessor>
      <resolver>org.greenplum.pxf.plugins.jdbc.JdbcResolver</resolver>
    </plugins>
    </profile>
</profiles>

``` 
 
</details>
2. create server configuration in $PXF_BASE/servers/postgresql/jdbc-site.xml
<details>
  <summary>Ex</summary>

 ```
<?xml version="1.0" encoding="UTF-8"?>
<configuration>
    <property>
        <name>jdbc.driver</name>
        <value>org.postgresql.Driver</value>
        <description>Class name of the JDBC driver (e.g. org.postgresql.Driver)</description>
    </property>
    <property>
        <name>jdbc.url</name>
        <value>jdbc:postgresql://localhost:5432/postgres</value>
        <description>The URL that the JDBC driver can use to connect to the database (e.g. jdbc:postgresql://localhost/postgres)</description>
    </property>
    <property>
        <name>jdbc.user</name>
        <value>bimbo</value>
        <description>User name for connecting to the database (e.g. postgres)</description>
    </property>
</configuration>

``` 
</details>

3.
```
create extension pxf;
CREATE readable EXTERNAL TABLE test_bytea(x bytea)
     LOCATION ('pxf://public.test_bytea?PROFILE=JDBC&SERVER=postgresql')
          FORMAT 'CUSTOM' (FORMATTER='pxfwritable_import');

select * from test_bytea where x = '\132greenplum\132';
select * from test_bytea where x is NULL;
select * from test_bytea where x in ('\132greenplum\132', 'salsmasdm');

CREATE readable EXTERNAL TABLE test_bytea_arr(x bytea[])
     LOCATION ('pxf://public.test_bytea_arr?PROFILE=JDBC&SERVER=postgresql')
          FORMAT 'CUSTOM' (FORMATTER='pxfwritable_import');
select * from test_bytea_arr where x = array['\132greenplum\132'::bytea];
```
With fdw:
1. Create configuration:
<details>
  <summary>Ex</summary>

 ```
CREATE SERVER postgresql_server
     FOREIGN DATA WRAPPER jdbc_pxf_fdw
          OPTIONS ( 
         jdbc_driver 'org.postgresql.Driver', 
         db_url 'jdbc:postgresql://localhost:5432/postgres',
          batch_size '10000',
         fetch_size '2000'
     );
     
     
 CREATE USER MAPPING FOR my_user
     SERVER postgresql_server
     OPTIONS ( user 'pg_user');

``` 
 
</details>

2. The same queries:
```
            CREATE FOREIGN TABLE test_jsonb_fdw (t json)
          SERVER postgresql
          OPTIONS ( resource 'public.test_jsonb');

select * from test_jsonb_fdw where t in ('{"foo": [true, "bar"], "tags": {"a": 1, "b": null}}'::jsonb, '{"foo": [true, "bar"], "tags": {"a": 1, "b": null}}'::jsonb);
```
And then just look at pxf-service.log, or [here](https://github.com/arenadata/pxf/blob/103d3008fa6a1f02ec4b2bc47e9ffe242822558b/external-table/src/pxfheaders.c#L211), or at tcpdump

### [Supported operators](https://github.com/arenadata/pxf/blob/pxf-6.x/external-table/src/pxffilters.c#L325-L338) for filter pushdown:

- IN(single_value)
- Comparison operators >,>=,=,etc
- IS NULL
- IN (several values)
- expr1 BOOL_OP expr2

### MANUAL REGRESSION TEST run
in regression directory run:
`make FDW_FilterPushDownTest FilterPushDownTest`

### CODE REVIEW

**Comment by red1452** [comment]
Few small corrections of the description:

1. Extend the list
2. function supported_array_type(Oid type) is added
3. a new macro
4. because GPDB 6 does not have

**Review by red1452 — APPROVED** [review]

**Comment by whitehawk** [comment]
> The tests haven't been added due to absence of working test
framework.

is there any way to use unit tests from `external-table\test`?

**Comment by whitehawk** [comment]
In the testing description:
```
CREATE readable EXTERNAL TABLE test_bytea_arr(x bytea[])
     LOCATION ('pxf://public.test_bytea_arr?PROFILE=JDBC&SERVER=postgresql')
          FORMAT 'CUSTOM' (FORMATTER='pxfwritable_import');
select * from test_bytea where x = array['\132greenplum\132'::bytea];
```
shouldn't it be `select * from test_bytea_arr`?

**Comment by whitehawk** [comment]
In the description there is a typo 'string costants' -> 'string constants'.

File: external-table/src/pxffilters.c [hunk 1]
```diff
@@ -1239,22 +1537,36 @@ list_const_to_str(Const *constval, StringInfo buf)
...
+					if (ARR_ELEMTYPE(arr) == BOOLOID){
+						if (*extval == 't')
+							appendStringInfo(interm_buf, "%s", TrueConstValue);
+						else if (*extval == 'f')
+							appendStringInfo(interm_buf, "%s", FalseConstValue);
+						else
+							appendStringInfo(interm_buf, "%s", NullConstValue);
+					}
+					else
+						appendStringInfo(interm_buf, "%s", extval);
```
Comment (whitehawk) [inline external-table/src/pxffilters.c]: It is recommended to place the shortest branch first.

**Comment by bimboterminator1** [comment]
>is there any way to use unit tests from `external-table\test`?

I could try, but the tests version lagged behind the pxf extension version and can't be even compiled. The tests should be reviewed and recovered if possible. They also depend on gpdb source code and existed in pxf code base since the extension were the part of gpcontrib;

**Comment by whitehawk** [comment]
> I could try, but the tests version lagged behind the pxf extension version and can't be even compiled. The tests should be reviewed and recovered if possible. They also depend on gpdb source code and existed in pxf code base since the extension were the part of gpcontrib;

Ok, let's drop it for now.

File: fdw/pxf_filter.c [hunk 2]
```diff
@@ -1309,6 +1596,20 @@ ListConstToStr(Const *constval, StringInfo buf)
 		case INT4ARRAYOID:
 		case INT8ARRAYOID:
 		case TEXTARRAYOID:
+		case BOOLARRAYOID:
+		case BYTEAARRAYOID:
+		case FLOAT4ARRAYOID:
+		case FLOAT8ARRAYOID:
+		case BPCHARARRAYOID:
+		case VARCHAROID:
```
Comment (RekGRpth) [inline fdw/pxf_filter.c]:
```suggestion
		case VARCHARARRAYOID:
```

File: external-table/src/pxffilters.c [hunk 3]
```diff
@@ -1171,6 +1451,11 @@ scalar_const_to_str(Const *constval, StringInfo buf)
 		case BYTEAOID:
 		case DATEOID:
 		case TIMESTAMPOID:
+		case TIMEOID:
+		case TIMESTAMPTZOID:
+		case INTERVALOID:
+		case UUIDOID:
+		case JSONBOID:
```
Comment (RekGRpth) [inline external-table/src/pxffilters.c:1549]: where is JSONOID?

File: external-table/src/pxffilters.c [hunk 4]
```diff
@@ -1214,6 +1498,20 @@ list_const_to_str(Const *constval, StringInfo buf)
...
+		case BPCHARARRAYOID:
+		case VARCHARARRAYOID:
+		case DATEARRAYOID:
+		case TIMEARRAYOID:
+		case TIMESTAMPARRAYOID:
+		case TIMESTAMPTZARRAYOID:
+		case INTERVALARRAYOID:
+		case NUMERICARRAYOID:
+		case UUIDARRAYOID:
+		case JSONBARRAYOID:
```
Comment (RekGRpth) [inline external-table/src/pxffilters.c:1611]: where is JSONARRAYOID?

File: external-table/src/pxffilters.c [hunk 5]
```diff
@@ -190,6 +208,58 @@ dbop_pxfop_map pxf_supported_opr_op_expr[] =
...
+	{2977 /* uuid_ge */ , PXFOP_GE},
+	{2973 /* uuid_ne */ , PXFOP_NE},
+
+	/* jsonb */
+	{3240 /* jsonb_eq */ , PXFOP_EQ},
+	{3242 /* jsonb_lt */ , PXFOP_LT},
+	{3243 /* jsonb_gt */ , PXFOP_GT},
+	{3244 /* jsonb_le */ , PXFOP_LE},
+	{3245 /* jsonb_ge */ , PXFOP_GE},
+	{3241 /* jsonb_ne */ , PXFOP_NE},
```
Comment (RekGRpth) [inline external-table/src/pxffilters.c:267]: whereis json?

File: external-table/src/pxffilters.c [hunk 6]
```diff
@@ -237,10 +307,86 @@ dbop_pxfop_array_map pxf_supported_opr_scalar_array_op_expr[] =
...
+
+	{TimestampTZEqualOperator /* time_eq */ , PXFOP_IN, true},
+
+	{IntervalEqualOperator /* interval_eq */ , PXFOP_IN, true},
+
+	{NumericEqualOperator /* numericeq */ , PXFOP_IN, true},
+
+	{UuidEqualOperator /* uuid_eq */ , PXFOP_IN, true},
+
+	{3240 /* jsonb_eq */ , PXFOP_IN, true},
```
Comment (RekGRpth) [inline external-table/src/pxffilters.c:362]: where is json?

File: external-table/src/pxffilters.c [hunk 7]
```diff
@@ -237,10 +307,86 @@ dbop_pxfop_array_map pxf_supported_opr_scalar_array_op_expr[] =
...
+#define NUMERICARRAYOID 1231
+#endif
+
+#ifndef UUIDARRAYOID
+#define UUIDARRAYOID 2951
+#endif
+
+#ifndef JSONBARRAYOID
+#define JSONBARRAYOID 3807
+#endif
```
Comment (RekGRpth) [inline external-table/src/pxffilters.c:415]: where is JSONARRAYOID?

**Comment by RekGRpth** [comment]
> > is there any way to use unit tests from `external-table\test`?
> 
> I could try, but the tests version lagged behind the pxf extension version and can't be even compiled. The tests should be reviewed and recovered if possible. They also depend on gpdb source code and existed in pxf code base since the extension were the part of gpcontrib;

how about to add tests like in https://github.com/arenadata/pxf/pull/21?

File: external-table/src/pxffilters.c (hunk 5 above)
Comment (bimboterminator1) [reply external-table/src/pxffilters.c:267]: Scalar operators are not supported for JSON type, only for jsonb. Thats why json is not everywhere. Is NULL is supported, other operators are not.
```
postgres=# select * from test_json where t = '{"foo": [true, "bar"], "tags": {"a": 1, "b": null}}'::json;
ERROR:  operator does not exist: json = json
```

**Comment by bimboterminator1** [comment]
> how about to add tests like in #21?

We will need to [extend java part ](https://github.com/arenadata/pxf/blob/pxf-6.x/server/pxf-diagnostic/src/main/java/org/greenplum/pxf/diagnostic/UserDataVerifyAccessor.java#L44-L52) (or refactor). And about regression tests the table definition will look like:
```
CREATE EXTERNAL TABLE test_filter (t0 text, a1 integer, b2 boolean, c3 numeric, d4 char(2), e5 varchar(2), x2 bytea,
    x9 date, x10 time, x11 timestamp, x12 timestamp with time zone,
    x13 interval, x14 uuid, x15 json, x16 jsonb, x17 int2[], x18 int4[], x19 int8[], x20 bool[], x21 text[],
    x22 float4[], x23 float8[], x24 bpchar[], x25 varchar(2)[], x26 date[], x27 uuid[], x28 numeric[],
    x29 time[], x30 timestamp[], x31 timestamp with time zone[], x32 interval[],
    x33 json[], x34 jsonb[], filterValue text)
    LOCATION (E'pxf://dummy_path?PROFILE=system:filter')
    FORMAT 'CSV';
    
```
Is it ok? At first glance it seems quite easy to edit java code

**Comment by RekGRpth** [comment]
> select * from test_bytea where x = '\132greenplum\132';

why
```sql
set client_min_messages = debug1;
select * from ee where x = '\132greenplum\132';
DEBUG1:  supported_array_type: filter pushdown is not supported for datatype oid: 17  (seg0 slice1 172.19.0.3:6002 pid=125148)
```
?

**Comment by RekGRpth** [comment]
> > select * from test_bytea where x = '\132greenplum\132';
> 
> why
> 
> ```sql
> set client_min_messages = debug1;
> select * from ee where x = '\132greenplum\132';
> DEBUG1:  supported_array_type: filter pushdown is not supported for datatype oid: 17  (seg0 slice1 172.19.0.3:6002 pid=125148)
> ```
> 
> ?

I meant, how good is it to call supported_array_type for non-array types?

**Comment by bimboterminator1** [comment]
> I meant, how good is it to call supported_array_type for non-array types?

I wanted to check somehow whether given Const is array or not. Because for the query like
```
CREATE readable EXTERNAL TABLE test_bytea_arr(x bytea[])
     LOCATION ('pxf://public.test_bytea_arr?PROFILE=JDBC&SERVER=postgresql')
          FORMAT 'CUSTOM' (FORMATTER='pxfwritable_import');
select * from test_bytea_arr where x = array['\000\132greenplum\000\132'::bytea, NULL];
```
the execution will get to the opexpr_to_pxffilter(). And i want to serialize array types via list_const_to_str(), not via scalar_const_to_str(). How good? I just compare oids, supported_array_type() is designed for the cases like above.

File: external-table/src/pxffilters.c [hunk 8]
```diff
@@ -611,7 +818,7 @@ pxf_serialize_filter_list(List *expressionItems)
 						else if (pxfoperand_is_list_const(l) && pxfoperand_is_attr(r))
 						{
 							appendStringInfo(resbuf, "%c%d%s%c%d",
-											 PXF_SCALAR_CONST_CODE, l.consttype,
+											 PXF_LIST_CONST_CODE, l.consttype,
```
Comment (RekGRpth) [inline external-table/src/pxffilters.c:860]: where is such change in fdw?

File: external-table/src/pxffilters.c [hunk 9]
```diff
@@ -257,11 +403,56 @@ Oid			pxf_supported_types[] =
...
+	VARCHARARRAYOID,
+	DATEARRAYOID,
+	TIMEARRAYOID,
+	TIMESTAMPARRAYOID,
+	TIMESTAMPTZARRAYOID,
+	INTERVALARRAYOID,
+	NUMERICARRAYOID,
+	UUIDARRAYOID,
+	JSONBARRAYOID,
+	JSONARRAYOID
```
Comment (RekGRpth) [inline external-table/src/pxffilters.c]:
```suggestion
	JSONARRAYOID,
```
like in fdw

File: external-table/src/pxffilters.c [hunk 10]
```diff
@@ -257,11 +403,56 @@ Oid			pxf_supported_types[] =
...
```
Comment (RekGRpth) [inline external-table/src/pxffilters.c]:
```suggestion
	JSONARRAYOID,
```
like in fdw

**Comment by bimboterminator1** [comment]
I've also tried to launch regression tests.  Queries for most of the types can be run within regress framework by applying the patch in the PR description.

**Comment by RekGRpth** [comment]
> ### MANUAL REGRESSION TEST run
> 
>     1. apply the patch:
>        [TESTPATCH.txt](https://github.com/user-attachments/files/16090978/TESTPATCH.txt)
> 
>     2. in regression directory run make 5772Test.
> 
>     3. See the output in result directory

I think the java part is quite simple to add these tests to the patch.

**Comment by bimboterminator1** [comment]
It's not the only part which should be modified. The existing java tests, regression tests will also be affected, the diff will be huge enough. And the tests anyway won't cover all the types mentioned in this patch. At least there is a bunch of errors for the timestamp with time zone, jsonb, and overcoming those issues is quite burdensome

**Comment by bimboterminator1** [comment]
Have added regression tests, just run `make FDW_FilterPushDownTest FilterPushDownTest` in regression directory.
I couldn't deal with automation tests and changed those files without extra checking.

File: regression/expected/FDW_FilterPushDownTest.out [hunk 11]
```diff
@@ -177,13 +217,13 @@ SELECT * FROM test_filter WHERE  t0 <> 'C' ORDER BY t0, a1;
...
 
-SELECT * FROM test_filter WHERE  t0 LIKE     'C%' ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  t0 LIKE     'C%' ORDER BY t0, a1;
  t0 | a1 | b2 |  c3  | d4 | e5 | filtervalue  
 ----+----+----+------+----+----+--------------
  C  |  2 |    | 2.21 | CC | CC | a0c25s2dC%o7
 (1 row)
 
-SELECT * FROM test_filter WHERE  t0 NOT LIKE 'C%' ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  t0 NOT LIKE 'C%' ORDER BY t0, a1;
```
Comment (RekGRpth) [inline regression/expected/FDW_FilterPushDownTest.out:226]: Why text NOT LIKE is not supported?

File: regression/expected/FDW_FilterPushDownTest.out [hunk 12]
```diff
@@ -196,14 +236,14 @@ SELECT * FROM test_filter WHERE  t0 NOT LIKE 'C%' ORDER BY t0, a1;
...
-SELECT * FROM test_filter WHERE  t0 IN     ('C','D') ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  t0 IN     ('C','D') ORDER BY t0, a1;
  t0 | a1 | b2 |  c3  | d4 | e5 |    filtervalue     
 ----+----+----+------+----+----+--------------------
  C  |  2 |    | 2.21 | CC | CC | a0m1009s1dCs1dDo10
  D  |  3 | f  |      | DD | DD | a0m1009s1dCs1dDo10
 (2 rows)
 
-SELECT * FROM test_filter WHERE  t0 NOT IN ('C','D') ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  t0 NOT IN ('C','D') ORDER BY t0, a1;
```
Comment (RekGRpth) [inline regression/expected/FDW_FilterPushDownTest.out:246]: Why text NOT IN is not supported?

File: regression/expected/FDW_FilterPushDownTest.out [hunk 13]
```diff
@@ -312,14 +352,14 @@ SELECT * FROM test_filter WHERE  a1 <> 2 ORDER BY t0, a1;
...
-SELECT * FROM test_filter WHERE  a1 IN     (2,3) ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  a1 IN     (2,3) ORDER BY t0, a1;
  t0 | a1 | b2 |  c3  | d4 | e5 |    filtervalue     
 ----+----+----+------+----+----+--------------------
  C  |  2 |    | 2.21 | CC | CC | a1m1007s1d2s1d3o10
  D  |  3 | f  |      | DD | DD | a1m1007s1d2s1d3o10
 (2 rows)
 
-SELECT * FROM test_filter WHERE  a1 NOT IN (2,3) ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  a1 NOT IN (2,3) ORDER BY t0, a1;
```
Comment (RekGRpth) [inline regression/expected/FDW_FilterPushDownTest.out:362]: Why integer NOT IN is not supported?

File: regression/expected/FDW_FilterPushDownTest.out [hunk 14]
```diff
@@ -428,14 +468,14 @@ SELECT * FROM test_filter WHERE  c3 <> 1.11 ORDER BY t0, a1;
...
- C  |  2 |    | 2.21 | CC | CC | No filter
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  c3 IN     (1.11,2.21) ORDER BY t0, a1;
+ t0 | a1 | b2 |  c3  | d4 | e5 |       filtervalue        
+----+----+----+------+----+----+--------------------------
+ B  |    | f  | 1.11 | BB | BB | a3m1231s4d1.11s4d2.21o10
+ C  |  2 |    | 2.21 | CC | CC | a3m1231s4d1.11s4d2.21o10
 (2 rows)
 
-SELECT * FROM test_filter WHERE  c3 NOT IN (1.11,2.21) ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  c3 NOT IN (1.11,2.21) ORDER BY t0, a1;
```
Comment (RekGRpth) [inline regression/expected/FDW_FilterPushDownTest.out:478]: Why numeric NOT IN is not supported?

File: regression/expected/FDW_FilterPushDownTest.out [hunk 15]
```diff
@@ -550,13 +590,13 @@ SELECT * FROM test_filter WHERE  d4 <> 'BB' ORDER BY t0, a1;
...
-----+----+----+------+----+----+-------------
- B  |    | f  | 1.11 | BB | BB | No filter
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  d4 LIKE     'B%' ORDER BY t0, a1;
+ t0 | a1 | b2 |  c3  | d4 | e5 | filtervalue  
+----+----+----+------+----+----+--------------
+ B  |    | f  | 1.11 | BB | BB | a4c25s2dB%o7
 (1 row)
 
-SELECT * FROM test_filter WHERE  d4 NOT LIKE 'B%' ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  d4 NOT LIKE 'B%' ORDER BY t0, a1;
```
Comment (RekGRpth) [inline regression/expected/FDW_FilterPushDownTest.out:599]: Why char NOT LIKE is not supported?

File: regression/expected/FDW_FilterPushDownTest.out [hunk 16]
```diff
@@ -569,14 +609,14 @@ SELECT * FROM test_filter WHERE  d4 NOT LIKE 'B%' ORDER BY t0, a1;
...
- C  |  2 |    | 2.21 | CC | CC | No filter
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  d4 IN     ('BB','CC') ORDER BY t0, a1;
+ t0 | a1 | b2 |  c3  | d4 | e5 |     filtervalue      
+----+----+----+------+----+----+----------------------
+ B  |    | f  | 1.11 | BB | BB | a4m1014s2dBBs2dCCo10
+ C  |  2 |    | 2.21 | CC | CC | a4m1014s2dBBs2dCCo10
 (2 rows)
 
-SELECT * FROM test_filter WHERE  d4 NOT IN ('BB','CC') ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  d4 NOT IN ('BB','CC') ORDER BY t0, a1;
```
Comment (RekGRpth) [inline regression/expected/FDW_FilterPushDownTest.out:619]: Why char NOT IN is not supported?

File: regression/expected/FDW_FilterPushDownTest.out [hunk 17]
```diff
@@ -690,13 +730,13 @@ SELECT * FROM test_filter WHERE  e5 <> 'BB' ORDER BY t0, a1;
...
 
-SELECT * FROM test_filter WHERE  e5 LIKE     'B%' ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  e5 LIKE     'B%' ORDER BY t0, a1;
  t0 | a1 | b2 |  c3  | d4 | e5 | filtervalue  
 ----+----+----+------+----+----+--------------
  B  |    | f  | 1.11 | BB | BB | a5c25s2dB%o7
 (1 row)
 
-SELECT * FROM test_filter WHERE  e5 NOT LIKE 'B%' ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  e5 NOT LIKE 'B%' ORDER BY t0, a1;
```
Comment (RekGRpth) [inline regression/expected/FDW_FilterPushDownTest.out:739]: Why varchar NOT LIKE is not supported?

File: regression/expected/FDW_FilterPushDownTest.out [hunk 18]
```diff
@@ -709,14 +749,14 @@ SELECT * FROM test_filter WHERE  e5 NOT LIKE 'B%' ORDER BY t0, a1;
...
- C  |  2 |    | 2.21 | CC | CC | No filter
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  e5 IN     ('BB','CC') ORDER BY t0, a1;
+ t0 | a1 | b2 |  c3  | d4 | e5 |     filtervalue      
+----+----+----+------+----+----+----------------------
+ B  |    | f  | 1.11 | BB | BB | a5m1009s2dBBs2dCCo10
+ C  |  2 |    | 2.21 | CC | CC | a5m1009s2dBBs2dCCo10
 (2 rows)
 
-SELECT * FROM test_filter WHERE  e5 NOT IN ('BB','CC') ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  e5 NOT IN ('BB','CC') ORDER BY t0, a1;
```
Comment (RekGRpth) [inline regression/expected/FDW_FilterPushDownTest.out:759]: Why varchar NOT IN is not supported?

File: regression/expected/FDW_FilterPushDownTest.out [hunk 19]
```diff
@@ -913,13 +953,13 @@ SELECT * FROM test_filter WHERE  t0 <> 'C' ORDER BY t0, a1;
...
 
-SELECT * FROM test_filter WHERE  t0 LIKE     'C%' ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  t0 LIKE     'C%' ORDER BY t0, a1;
  t0 | a1 | b2 |  c3  | d4 | e5 | filtervalue  
 ----+----+----+------+----+----+--------------
  C  |  2 |    | 2.21 | CC | CC | a0c25s2dC%o7
 (1 row)
 
-SELECT * FROM test_filter WHERE  t0 NOT LIKE 'C%' ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE  t0 NOT LIKE 'C%' ORDER BY t0, a1;
```
Comment (RekGRpth) [inline regression/expected/FDW_FilterPushDownTest.out:962]: Why text NOT LIKE is not supported?

File: regression/expected/FDW_FilterPushDownTest.out (hunk 11 above)
Comment (bimboterminator1) [reply regression/expected/FDW_FilterPushDownTest.out:226]: The main task is to extend the supported type list, supporting new operators is a separate issue.

File: automation/sqlrepo/features/filterpushdown/checkFilterPushDown/expected/query01.ans [hunk 20]
```diff
@@ -188,13 +188,13 @@ SELECT * FROM test_filter WHERE t0 <> 'C' ORDER BY t0, a1;
...
 
-SELECT * FROM test_filter WHERE t0 LIKE     'C%' ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE t0 LIKE     'C%' ORDER BY t0, a1;
  t0 | a1 | b2 |  c3  | d4  | e5 | filtervalue
 ----+----+----+------+-----+----+--------------
  C  |  2 |    | 2.21 | CC  | CC | a0c25s2dC%o7
 (1 row)
 
-SELECT * FROM test_filter WHERE t0 NOT LIKE 'C%' ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE t0 NOT LIKE 'C%' ORDER BY t0, a1;
```
Comment (RekGRpth) [inline automation/sqlrepo/features/filterpushdown/checkFilterPushDown/expected/query01.ans:197]: I don’t understand why there is no filter here, but in the same query below in FDW_FilterPushDownTest.out there is one?

Comment (bimboterminator1) [reply automation/sqlrepo/features/filterpushdown/checkFilterPushDown/expected/query01.ans:197]: i haven't run this test suite. Therefore i make changes without checking. Thank you for your careful review.

File: automation/sqlrepo/features/filterpushdown/checkFilterPushDown/expected/query01.ans [hunk 21]
```diff
@@ -439,34 +439,34 @@ SELECT * FROM test_filter WHERE c3 <> 1.11 ORDER BY t0, a1;
     |  0 | t  | 0.01 | AA  | AA | a3c1700s4d1.11o6
 (8 rows)
 
-SELECT * FROM test_filter WHERE c3 IN     (1.11,2.21) ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE c3 IN     (1.11,2.21) ORDER BY t0, a1;
  t0 | a1 | b2 |  c3  | d4  | e5 | filtervalue
 ----+----+----+------+-----+----+-------------
  B  |    | f  | 1.11 | BB  | BB | No filter
```
Comment (RekGRpth) [inline automation/sqlrepo/features/filterpushdown/checkFilterPushDown/expected/query01.ans]: still not supported?

File: automation/sqlrepo/features/filterpushdown/checkFilterPushDown/expected/query02.ans [hunk 22]
```diff
@@ -701,53 +701,53 @@ SELECT * FROM test_filter WHERE e5 <> 'BB' ORDER BY t0, a1;
...
+ I  |  8 | t  | 8.81 | II | II | a5c25s2dB%o7l2
+ J  |  9 | f  | 9.91 | JJ | JJ | a5c25s2dB%o7l2
+    |  0 | t  | 0.01 | AA | AA | a5c25s2dB%o7l2
 (8 rows)
 
-SELECT * FROM test_filter WHERE e5 IN     ('BB','CC') ORDER BY t0, a1;
+SELECT t0, a1, b2, c3, d4, e5, filterValue FROM test_filter WHERE e5 IN     ('BB','CC') ORDER BY t0, a1;
  t0 | a1 | b2 |  c3  | d4  | e5 | filtervalue 
 ----+----+----+------+-----+----+-------------
  B  |    | f  | 1.11 | BB  | BB | No filter
```
Comment (RekGRpth) [inline automation/sqlrepo/features/filterpushdown/checkFilterPushDown/expected/query02.ans]: still not supported?

File: fdw/pxf_filter.c [hunk 23]
```diff
@@ -192,65 +212,221 @@ dbop_pxfop_map pxf_supported_opr_op_expr[] =
...
+	{1321 /* timestamptz_ne */ , PXFOP_NE},
+
+	/* interval */
+	{IntervalEqualOperator /* interval_eq */ , PXFOP_EQ},
+	{1332 /* interval_lt */ , PXFOP_LT},
+	{1334 /* interval_gt */ , PXFOP_GT},
+	{1333 /* interval_le */ , PXFOP_LE},
+	{1335 /* interval_ge */ , PXFOP_GE},
+	{1331 /* interval_ne */ , PXFOP_NE},
+
```
Comment (RekGRpth) [inline fdw/pxf_filter.c:254]:
```suggestion

	/* uuid */
```
(like in external table)

File: fdw/pxf_filter.c [hunk 24]
```diff
@@ -192,65 +212,221 @@ dbop_pxfop_map pxf_supported_opr_op_expr[] =
...
+	{1753 /* numericne */ , PXFOP_NOTIN, false},
+
+	{UuidEqualOperator /* uuid_eq */ , PXFOP_IN, true},
+	{2973 /* uuid_ne */ , PXFOP_NOTIN, false},
+
+	{3240 /* jsonb_eq */ , PXFOP_IN, true},
+	{3241 /* jsonb_ne */ , PXFOP_NOTIN, false},
 };
 
+
```
Comment (RekGRpth) [inline fdw/pxf_filter.c]:
```suggestion
```
(like in external table)

**Comment by RekGRpth** [comment]
Files checkFilterPushDown/sql/query01.sql and checkFilterPushDown/sql/query02.sql differ only in the explicit setting of the optimizer, so I think that the outputs for them should differ only in this.
Also I think that filtervalue should be same in regression/expected/FDW_FilterPushDownTest.out and regression/expected/FilterPushDownTest.out too.

**Review by RekGRpth — APPROVED** [review]

**Pull request merged**